### DISCUSSION TITLE
rio.Map: An Advanced Map Component Using

### DISCUSSION THREAD

**Discussion by JJFlorian** [rio-labs/rio#29] [Ideas]
Currently, if you want to add a map to your website, using Plotly in a rio.Plot component is the way to go. However, Plotly is not the only available option to visualize geospatial data. Many engines exist, such as Bokeh, Deck.gl, Folium, Kepler.gl, and many more.

So far, I have been using Deck.gl and Folium, but I would like to hear what other people use and why. If there seems to be a best option, that would, of course, be the solution. However, if possible on the Rio side, it would be nice to allow the developer to choose the engine of their liking.

Leafmap is a nice package that is built on top of all these optional engines. I can't tell whether it can be used for a Map component. That's up to the core developers.

**Comment by JJFlorian** [comment]
Hey Rio devs,

Have you had a discussion on the map component yet? The Plotly map example looks nice and covers a lot of geospatial visualization needs. However, I think many developers wouldn’t prefer Plotly for maps and would appreciate alternative options in Rio.

I've explored map options and created a custom Map component using pure Leaflet on the frontend. It works, but lacks the features of established packages. I understand creating a custom map API might not be ideal.

The main issue with the packages I've looked into is the lack of an easy frontend translation like Plotly's `to_json` method on the backend and `newPlot` on the frontend.

However, many Map objects in the Leafmap package have a `to_html` method. Do you think it’s feasible and worthwhile to send these HTML strings to the frontend and parse them to dynamically create the DOM in the Map component? If not, what alternatives do you suggest? Contributing to the Python map packages, by for example developing a similar to_json method, seems less useful without a way to transform JSON data into a map on the frontend.

**Comment by mad-moo** [comment]
At the end of the day, Rio components are just HTML. Any engine that can display their map using JavaScript/HTML should be possible to use. I think the question here is more about the map API itself. I'm sure the second a map component is added, a lot of related requests will pop up, like how one can add pins. How to display a fictional map, ...

I'd love to hear input from somebody who has experience with different mapping engines and has already implemented a variety of uses cases with them, so we can hear about commonly needed features.

**Reply by JJFlorian** [reply]
Ok great.

My vote will definetely go to [folium](https://github.com/python-visualization/folium). It pretty much covers everything. On the visualisation side it covers raster data and vector data. Besides it has nice UI element features, allowing the user to add custom icons, overlays (guess you should use these or WMS layers for fictional maps), splitted maps, and much more. On the data intake side it covers the convential formats such as geojson and geopandas. 

Optionally the [foliumap](https://github.com/opengeos/leafmap/blob/master/leafmap/foliumap.py) module within leafmap could be chosen to add even more options. It enables even more visualisations such as custom legends and a selection of basemaps. It also expands the data input options such as kml and google earth engine. 

I hope to see some other suggestions and opinions in this discussion!

**Comment by ivanthewebber** [comment]
It's worth stating explicitly that maps introduce certain localization challenges (e.g. showing disputed borders), generally need geocoding for any user input, and would require licensing for map data and satellite/aerial imagery. Thus, it makes sense to build map components using existing map modules and services.

There are many map modules that can be used to build a map component. These have some differences in how they look and the options they provide. Perhaps we should start by collecting links to existing map component implementations (@JJFlorian I assume you made yours). 

Maps have multiple layers (e.g. roads, buildings, businesses) which are like tables of data with a geospatial geometry. A base map provides some of those layers, but often they are augmented with one or more subject-matter layers. Like a table (and often displayed in table form as well) these layers often have filters. Thus, I think a good map component design would build on Rio's Table. Ideally columns could be used for coloring (clustering) or filtering. I think Rio should define a MapComponent abstract base class (with perhaps a very simple implementation) that can then be extended for different map providers while maintaining the same UI/table-sync.

I plan to take a swing at synced map & table and make something like [the site in this webinar](https://www.mapbox.com/webinars/mapbox-python-anvil).

**Reply by ivanthewebber** [reply]
This also seems useful for making map plots more interactive: https://medium.com/data-science/how-to-create-interactive-map-plots-with-plotly-7b57e889239a.

**Reply by ivanthewebber** [reply]
[Reflex's map](https://reflex.dev/docs/enterprise/map/) is probably also a great comparison point
